node/src/components/linear_chain.rs

```diff
@@ -444,22 +461,32 @@ where
                     self.signature_cache.insert(*signatures.clone());
                 }
                 // Check if the validator is bonded in the era in which the block was created.
-                effect_builder
-                    .is_bonded_validator(fs.era_id, fs.public_key)
-                    .map(|is_bonded| {
-                        if is_bonded {
-                            Ok((maybe_signatures, fs, is_bonded))
-                        } else {
-                            Err((maybe_signatures, fs))
-                        }
-                    })
+                // TODO: Use protocol version that is valid for the block's height.
+                let protocol_version = self.protocol_version;
+                let initial_state_root_hash = self.initial_state_root_hash;
+                let activation_era_id = self.activation_era_id;
+                async move {
+                    match effect_builder
+                        .is_bonded_validator(
+                            &fs.public_key,
+                            fs.era_id,
+                            activation_era_id,
+                            initial_state_root_hash,
+                            protocol_version,
+                        )
+                        .await
+                    {
+                        Some(is_bonded) => Ok((maybe_signatures, fs, is_bonded)),
+                        None => Err((maybe_signatures, fs)),
+                    }
+                }
+                .result(
+                    |(maybe_signatures, fs, is_bonded)| {
+                        Event::IsBonded(maybe_signatures, fs, is_bonded)
+                    },
+                    |(maybe_signatures, fs)| Event::IsBondedFutureEra(maybe_signatures, fs),
+                )
```

Comment:
Rather than call `.result` here, can't we move this into the `Some`/`None` match case above?

Reply:
IIRC, it won't type check.

Reply:
I'm not sure - the problem is, I think effects aren't supposed to return events directly? They usually return some value, which is then wrapped in an event, like here. But maybe it can be done somehow? I'm open to suggestions.

Reply:
It's pretty confusing to have two match arms in a row.

However we can do this:

```rust
                // Check if the validator is bonded in the era in which the block was created.
                effect_builder
                    .is_bonded_validator(fs.era_id, fs.public_key)
                    .map(|is_bonded| {
                        if is_bonded {
                            Event::IsBonded(maybe_signatures, fs, is_bonded)
                        } else {
                            Event::IsBondedFutureEra(maybe_signatures, fs)
                        }
                    })
                    .event(std::convert::identity)
```

Reply:
Nice. Though with the latest version, there is no `IsBondedFutureEra` anymore, so it became a bit simpler (at the cost of making the effect more complex :grimacing: ).